### ISSUE TITLE
null flow state should be shown at declaration site as well as use site

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#42897]
**Version Used**: VS 16.6 preview 1

**Steps to Reproduce**:

In VS copy the following:

```csharp
#nullable enable
public class C
{
    string M()
    {
        var str = "hello";
        return str;
    }
}
```

Hover over `str` in `return str`.

You get the following popup:

![image](https://user-images.githubusercontent.com/29174528/77884120-253d5680-726d-11ea-8c1a-c11f2a9248c5.png)

Hover over `str` in `var str = "hello";`

You get the following popup:

![image](https://user-images.githubusercontent.com/29174528/77884206-4aca6000-726d-11ea-82c8-b4c58703f525.png)

Note it doesn't mention the null flow state, only the null-type of the variable.

I found this extremely confusing since `str` is very clearly not null. I wasted a few minutes trying to find out if the method I was calling returned null, before I realized it was because `var` is always inferred as nullable.

I think this would be less confusing if the null flow state was also shown in declaration so it would say:
> (local variable) string? str 
> str is not null here"

**Comment by CyrusNajmabadi** [comment]
@jcouv i agree.  This is def a bit confusing.  Hopefully this is a low-cost thing to implement.

**Comment by sharwell** [comment]
We need to think about the way this feature interacts with pre- and post-conditions.

* Some cases have the same pre- and post-conditions (e.g. most reads)
* Some cases only have post-conditions (e.g. variable initializers)
* Some cases have both pre- and post-conditions, e.g. the flow state of `value` here:

    ```csharp
    Debug.Assert(value is object);
    ```

    For this expression, the post-condition is always "not null", but the precondition could be any of "not null", "maybe null", or "unknown (oblivious)".

**Comment by sharwell** [comment]
Design review conclusion: (all examples here show the not-null case)

### Overview

By default, show the following:

> 'str' is not null here.

If the post-condition of the use changes the flow analysis state, show the following:

> 'str' might be null here.
> 'str' is not null after this use.

If the use is the target of an assignment (where the value is written but not read):

> 'str' is not null after this assignment.

### Special cases

* Only show the pre- and post-conditions if the use is not just an assignment _and_ the flow analysis state changes as a result of the use.
